ci/Dockerfile

```
@@ -31,80 +35,89 @@ RUN set -eux; \
+    apt-get autoremove -y && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* /usr/share/doc /usr/share/man /usr/local/doc /usr/lib/llvm-11 /tmp/* /var/tmp/*
 
 ### generic ci ####
 
 ARG GH_TOOL_VERSION="2.54.0"
 
-RUN wget "https://github.com/cli/cli/releases/download/v${GH_TOOL_VERSION}/gh_${GH_TOOL_VERSION}_linux_amd64.deb" && \
-    dpkg -i "gh_${GH_TOOL_VERSION}_linux_amd64.deb" && rm "gh_${GH_TOOL_VERSION}_linux_amd64.deb"
+RUN wget "https://github.com/cli/cli/releases/download/v${GH_TOOL_VERSION}/gh_${GH_TOOL_VERSION}_linux_${TARGETARCH}.deb" && \
+    dpkg -i "gh_${GH_TOOL_VERSION}_linux_${TARGETARCH}.deb" && \
+    rm "gh_${GH_TOOL_VERSION}_linux_${TARGETARCH}.deb"
 
-# generic ci | install stable rust
+# generic ci | install rust toolchains
 # llvm-tools-preview is for grcov
 # base | install rustup, use minimum components
-RUN curl -L "https://static.rust-lang.org/rustup/dist/x86_64-unknown-linux-gnu/rustup-init" \
-    -o rustup-init && \
+RUN case ${TARGETARCH} in \
+        "amd64")  RUST_ARCH="x86_64"  ;; \
+        "arm64")  RUST_ARCH="aarch64" ;; \
+        *)        RUST_ARCH=${TARGETARCH} ;; \
+    esac && \
+    curl -L "https://static.rust-lang.org/rustup/dist/${RUST_ARCH}-unknown-linux-gnu/rustup-init" -o rustup-init && \
     chmod u+x rustup-init && \
     ./rustup-init -y --no-modify-path --default-toolchain none && \
     rm -f rustup-init && \
     chown -R root:nonroot ${RUSTUP_HOME} ${CARGO_HOME} && \
     chmod -R g+w ${RUSTUP_HOME} ${CARGO_HOME} && \
+    # generic ci | install specific stable version
     rustup toolchain install "${RUST_STABLE_VERSION}" --profile minimal \
       --component rustfmt,clippy,rust-src,llvm-tools-preview && \
     rustup default "${RUST_STABLE_VERSION}" && \
     rm -rf "${CARGO_HOME}/registry" "${CARGO_HOME}/git" && \
     # generic ci | "alias" pinned stable toolchain as generic stable
-    ln -s "/usr/local/rustup/toolchains/${RUST_STABLE_VERSION}-x86_64-unknown-linux-gnu" /usr/local/rustup/toolchains/stable-x86_64-unknown-linux-gnu && \
-    # generic ci | install asm toolchain for the default stable toolchain
-    rustup target add riscv64imac-unknown-none-elf && \
-    # needed for `ink-node`
-    rustup target add wasm32v1-none && \
-    cargo install --git https://github.com/use-ink/ink-node --branch main --force --locked && \
+    ln -s "/usr/local/rustup/toolchains/${RUST_STABLE_VERSION}-${RUST_ARCH}-unknown-linux-gnu" "/usr/local/rustup/toolchains/stable-${RUST_ARCH}-unknown-linux-gnu" && \
+    # generic ci | install asm toolchain for the default stable toolchain; `wasm32v1-none` for `ink-node`
+    rustup target add riscv64imac-unknown-none-elf wasm32v1-none && \
     # generic ci | install specific rust nightly, default is stable, use minimum components
     rustup toolchain install "nightly-${RUST_NIGHTLY_VERSION}" --profile minimal \
       --component rustfmt,clippy,rust-analyzer,llvm-tools,rustc-dev,rust-src,llvm-tools-preview && \
```

```suggestion
      --component rustfmt,clippy,rust-analyzer,rustc-dev,rust-src,llvm-tools-preview && \
```
`llvm-tools` is just an alias/rename to `llvm-tools-preview`, so we only need one.
There's actually no real consensus on which name is preferrable, but I personally tend to use the `-preview` one (see https://github.com/rust-lang/rustup/pull/3578)